velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp

```diff
@@ -13,89 +13,214 @@
       auto group = groups[i];
-      comparisonValue(group) = initialValue_;
       valueIsNull(group) = true;
+
+      if (!isNumeric<T>()) {
+        new (group + offset_) SingleValueAccumulator();
+      }
+
+      if constexpr (isNumeric<U>()) {
+        *comparisonValue(group) = initialValue_;
+      } else {
+        new (
+            group + offset_ +
+            (isNumeric<T>() ? sizeof(T) : sizeof(SingleValueAccumulator)))
+            SingleValueAccumulator();
+      }
     }
   }
 
   void finalize(char** /* unused */, int32_t /* unused */) override {}
 
   void extractValues(char** groups, int32_t numGroups, VectorPtr* result)
       override {
-    auto vector = (*result)->as<FlatVector<T>>();
-    VELOX_CHECK(vector);
-    vector->resize(numGroups);
-    uint64_t* rawNulls = getRawNulls(vector);
+    VELOX_CHECK(result);
+    (*result)->resize(numGroups);
+    uint64_t* rawNulls = getRawNulls(result->get());
+
+    T* rawValues = nullptr;
+    if constexpr (isNumeric<T>()) {
+      auto vector = (*result)->as<FlatVector<T>>();
+      VELOX_CHECK(vector != nullptr);
+      rawValues = vector->mutableRawValues();
+    }
 
-    T* rawValues = vector->mutableRawValues();
     for (int32_t i = 0; i < numGroups; ++i) {
       char* group = groups[i];
       if (isNull(group) || valueIsNull(group)) {
-        vector->setNull(i, true);
+        (*result)->setNull(i, true);
       } else {
         clearNull(rawNulls, i);
-        rawValues[i] = this->value(group);
+        extract<T, ValueAccumulatorType>(value(group), *result, i, rawValues);
       }
     }
   }
 
   void extractAccumulators(char** groups, int32_t numGroups, VectorPtr* result)
       override {
     auto rowVector = (*result)->as<RowVector>();
-    auto valueVector = rowVector->childAt(0)->asFlatVector<T>();
-    auto comparisonVector = rowVector->childAt(1)->asFlatVector<U>();
+    auto valueVector = rowVector->childAt(0);
+    auto comparisonVector = rowVector->childAt(1);
 
     rowVector->resize(numGroups);
     valueVector->resize(numGroups);
     comparisonVector->resize(numGroups);
     uint64_t* rawNulls = getRawNulls(rowVector);
 
-    T* rawValues = valueVector->mutableRawValues();
-    U* rawComparisonValues = comparisonVector->mutableRawValues();
+    T* rawValues = nullptr;
+    if constexpr (isNumeric<T>()) {
+      auto flatValueVector = valueVector->as<FlatVector<T>>();
+      VELOX_CHECK(flatValueVector != nullptr);
+      rawValues = flatValueVector->mutableRawValues();
+    }
+    U* rawComparisonValues = nullptr;
+    if constexpr (isNumeric<U>()) {
+      auto flatComparisonVector = comparisonVector->as<FlatVector<U>>();
+      VELOX_CHECK(flatComparisonVector != nullptr);
+      rawComparisonValues = flatComparisonVector->mutableRawValues();
+    }
     BufferPtr nulls = valueVector->mutableNulls(rowVector->size());
-    uint64_t* nullValues = nulls->asMutable<uint64_t>();
+    uint64_t* rawValueNulls = nulls->asMutable<uint64_t>();
```

Comment:
nit: rawValueNulls -> rawNulls (to match 'nulls')

Reply:
rawNulls is defined for row vector at l183. I combine the two lines l197 and l198